### ISSUE TITLE
RyuJIT x86: SharpChess benchmark crashes

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#7113]
The SharpChess benchmark is crashing when run with RyuJIT/x86 at the following stack:

```
25 00efd930 5c6e3022 clr!DbgAssertDialog+0x2a8 [c:\dev\projectk\src\ndp\clr\src\utilcode\debug.cpp @ 745]
26 00efd9b4 5c6e318f clr!Frame::Push+0xe7 [c:\dev\projectk\src\ndp\clr\src\vm\frames.cpp @ 441]
27 00efda44 5cb44dae clr!HelperMethodFrame::Push+0xbc [c:\dev\projectk\src\ndp\clr\src\vm\frames.cpp @ 1732]
*** WARNING: Unable to verify checksum for C:\Windows\Microsoft.NET\Framework\v4.0.x86chk\assembly\NativeImages_v4.0.x86chk_32\mscorlib\e971c39439f5efb9a410089f31ab627a\mscorlib.ni.dll
28 (Inline) -------- clr!FrameWithCookie<HelperMethodFrame_1OBJ>::Push+0xb [c:\dev\projectk\src\ndp\clr\src\vm\frames.h @ 3627]
29 00efddec 57f024e8 clr!MarshalNative::SizeOfClass+0x298 [c:\dev\projectk\src\ndp\clr\src\vm\marshalnative.cpp @ 278]
2a 00efde00 06173cda mscorlib_ni!System.Runtime.InteropServices.Marshal.SizeOf(System.Type)+0x4c [C:\dev\ProjectK\src\ndp\clr\src\mscorlib\src\System\Runtime\InteropServices\Marshal.cs @ 253]
2b 00efde24 0617333d System_Windows_Forms_5b20000!System.Windows.Forms.Cursor.GetIconSize(IntPtr)+0x72 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Cursor.cs @ 597]
2c 00efde78 06173258 System_Windows_Forms_5b20000!System.Windows.Forms.Cursor.LoadPicture(IStream)+0xc5 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Cursor.cs @ 636]
2d 00efde8c 06172ea6 System_Windows_Forms_5b20000!System.Windows.Forms.Cursor..ctor(System.IO.Stream)+0x98 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Cursor.cs @ 154]
2e 00efdefc 0991d81a image00af0000!SharpChess.frmMain.LoadCursors()+0xc6
2f 00efdf0c 0991c6b0 image00af0000!SharpChess.frmMain.frmMain_Load(System.Object, System.EventArgs)+0x252
30 00efdf3c 0991bbcf System_Windows_Forms_5b20000!System.Windows.Forms.Form.OnLoad(System.EventArgs)+0x158 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Form.cs @ 4890]
31 00efdf48 0303ed28 System_Windows_Forms_5b20000!System.Windows.Forms.Form.OnCreateControl()+0x4f [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Form.cs @ 4696]
32 00efdf84 0303eb6c System_Windows_Forms_5b20000!System.Windows.Forms.Control.CreateControl(Boolean)+0x160 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 5793]
33 00efdf98 09913b09 System_Windows_Forms_5b20000!System.Windows.Forms.Control.CreateControl()+0x24 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 5733]
34 00efdfd0 05832800 System_Windows_Forms_5b20000!System.Windows.Forms.Control.WmShowWindow(System.Windows.Forms.Message ByRef)+0x91 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 13587]
35 00efe010 05b0f0aa System_Windows_Forms_5b20000!System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message ByRef)+0x7c0 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 13992]
36 00efe01c 05b0f038 System_Windows_Forms_5b20000!System.Windows.Forms.ScrollableControl.WndProc(System.Windows.Forms.Message ByRef)+0x5a [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\ScrollableControl.cs @ 1493]
37 00efe024 09913a64 System_Windows_Forms_5b20000!System.Windows.Forms.ContainerControl.WndProc(System.Windows.Forms.Message ByRef)+0x18 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\ContainerControl.cs @ 1905]
38 00efe034 05b0ed7f System_Windows_Forms_5b20000!System.Windows.Forms.Form.WmShowWindow(System.Windows.Forms.Message ByRef)+0x2c [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Form.cs @ 7294]
39 00efe048 05831c29 System_Windows_Forms_5b20000!System.Windows.Forms.Form.WndProc(System.Windows.Forms.Message ByRef)+0x2bf [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Form.cs @ 7480]
3a 00efe050 05831c03 System_Windows_Forms_5b20000!System.Windows.Forms.Control+ControlNativeWindow.OnMessage(System.Windows.Forms.Message ByRef)+0x11 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 14231]
3b 00efe064 058319bd System_Windows_Forms_5b20000!System.Windows.Forms.Control+ControlNativeWindow.WndProc(System.Windows.Forms.Message ByRef)+0xa3 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 14286]
3c 00efe0a0 0150d0de System_Windows_Forms_5b20000!System.Windows.Forms.NativeWindow.Callback(IntPtr, Int32, IntPtr, IntPtr)+0x65 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\NativeWindow.cs @ 650]
3d 00efe0d4 752b84f3 CLRStub[StubLinkStub]@a2f4950b0150d0de
3e 00efe100 75296c40 USER32!_InternalCallWinProc+0x2b [d:\th\windows\core\ntuser\client\i386\callproc.asm @ 116]
3f (Inline) -------- USER32!InternalCallWinProc+0x1b [d:\th\windows\core\ntuser\client\callproc.h @ 81]
40 00efe1a8 75296820 USER32!UserCallWinProcCheckWow+0x1f0 [d:\th\windows\core\ntuser\client\clmsg.c @ 184]
41 00efe208 7529d169 USER32!DispatchClientMessage+0xf0 [d:\th\windows\core\ntuser\client\client.c @ 3332]
42 00efe248 77e28e56 USER32!__fnDWORD+0x49 [d:\th\windows\core\ntuser\inc\ntcb.h @ 917]
43 00efe280 099139ee ntdll!KiUserCallbackDispatcher+0x36 [d:\th\minkernel\ntos\rtl\i386\userdisp.asm @ 517]
44 00efe2bc 0303c783 System_Windows_Forms_5b20000!DomainBoundILStubClass.IL_STUB_PInvoke(System.Runtime.InteropServices.HandleRef, Int32)+0x46
45 00efe35c 05b0e4d1 System_Windows_Forms_5b20000!System.Windows.Forms.Control.SetVisibleCore(Boolean)+0x103 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 11856]
46 00efe388 05b076a9 System_Windows_Forms_5b20000!System.Windows.Forms.Form.SetVisibleCore(Boolean)+0x181 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Form.cs @ 2676]
47 00efe390 05b0de62 System_Windows_Forms_5b20000!System.Windows.Forms.Control.set_Visible(Boolean)+0x11 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Control.cs @ 4292]
48 00efe3d8 05b0d920 System_Windows_Forms_5b20000!System.Windows.Forms.Application+ThreadContext.RunMessageLoopInner(Int32, System.Windows.Forms.ApplicationContext)+0xfa [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Application.cs @ 3352]
49 00efe408 05b0d804 System_Windows_Forms_5b20000!System.Windows.Forms.Application+ThreadContext.RunMessageLoop(Int32, System.Windows.Forms.ApplicationContext)+0x40 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Application.cs @ 3306]
4a 00efe420 030349aa System_Windows_Forms_5b20000!System.Windows.Forms.Application.Run(System.Windows.Forms.Form)+0x34 [C:\dev\ProjectK\src\NDP\fx\src\winforms\Managed\System\WinForms\Application.cs @ 1496]
4b 00efe42c 5c410bd1 image00af0000!SharpChess.frmMain.Main(System.String[])+0x4a
4c 00efe438 5c6e3acc clr!CallDescrWorkerInternal+0x34 [C:\dev\ProjectK\src\ndp\clr\src\vm\i386\asmhelpers.asm @ 787]
4d 00efe854 5c6e3cca clr!CallDescrWorker+0xdb [c:\dev\projectk\src\ndp\clr\src\vm\callhelpers.cpp @ 147]
4e 00efe940 5c6e4658 clr!CallDescrWorkerWithHandler+0x199 [c:\dev\projectk\src\ndp\clr\src\vm\callhelpers.cpp @ 89]
4f 00efec8c 5c6ca6ef clr!MethodDescCallSite::CallTargetWorker+0x8d5 [c:\dev\projectk\src\ndp\clr\src\vm\callhelpers.cpp @ 652]
50 (Inline) -------- clr!MethodDescCallSite::Call+0x11 [c:\dev\projectk\src\ndp\clr\src\vm\callhelpers.h @ 433]
51 00efed84 5c6ca4e8 clr!``RunMain'::`26'::__Body::Run'::`31'::__Body::Run+0x19c [c:\dev\projectk\src\ndp\clr\src\vm\assembly.cpp @ 2639]
52 00efeddc 5c6ca882 clr!`RunMain'::`26'::__Body::Run+0x71 [c:\dev\projectk\src\ndp\clr\src\vm\assembly.cpp @ 2603]
53 00efee3c 5c6c1a63 clr!RunMain+0xf1 [c:\dev\projectk\src\ndp\clr\src\vm\assembly.cpp @ 2655]
54 00eff200 5c4eb6c7 clr!Assembly::ExecuteMainMethod+0x1ea [c:\dev\projectk\src\ndp\clr\src\vm\assembly.cpp @ 2762]
55 00eff894 5c4059fa clr!SystemDomain::ExecuteMainMethod+0x836 [c:\dev\projectk\src\ndp\clr\src\vm\appdomain.cpp @ 3689]
56 00eff8dc 5c4052cb clr!``ExecuteEXE'::`10'::__Body::Run'::`15'::__Body::Run+0x5c [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 3077]
57 00eff934 5c3fb49e clr!`ExecuteEXE'::`10'::__Body::Run+0x71 [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 3072]
58 00eff974 5c408567 clr!ExecuteEXE+0x69 [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 3077]
59 00effa7c 5c40502c clr!_CorExeMainInternal+0x166 [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 2909]
5a 00effabc 5c40838c clr!`_CorExeMain'::`3'::__Body::Run+0x55 [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 2837]
5b 00effb00 5edac3bd clr!_CorExeMain+0x28 [c:\dev\projectk\src\ndp\clr\src\vm\ceemain.cpp @ 2837]
5c 00effb40 747edd05 mscoreei!_CorExeMain+0x13a [c:\dev\projectk\src\ndp\clr\src\dlls\shim\shim.cpp @ 6421]
5d (Inline) -------- MSCOREE!ShellShim__CorExeMain+0x94 [d:\th\com\netfx\windowsbuilt\shell_shim\v2api.cpp @ 284]
5e 00effb50 755138f4 MSCOREE!_CorExeMain_Exported+0xa5 [d:\th\com\netfx\windowsbuilt\shell_shim\v2api.cpp @ 1222]
5f 00effb64 77e15de3 KERNEL32!BaseThreadInitThunk+0x24 [d:\th\base\win32\client\thread.c @ 64]
60 00effbac 77e15dae ntdll!__RtlUserThreadStart+0x2f [d:\th\minkernel\ntdll\rtlstrt.c @ 997]
61 00effbbc 00000000 ntdll!_RtlUserThreadStart+0x1b [d:\th\minkernel\ntdll\rtlstrt.c @ 914]
```

It would seem that some function on the stack has corrupted the frame chain. This does not repro with JIT32.

**Comment by mikedn** [comment]
> It would seem that some function on the stack has corrupted the frame chain

I suppose this could be caused by #5699.

**Comment by pgavlin** [comment]
@mikedn: I probably should have been more specific: the frame chain I am referring to is not the ebp-based frame chain, but the linked list of `Frame` objects established for transitions between managed and native code. My guess is that something in PInvoke lowering isn't quite right for x86, but I'm not certain.

**Comment by pgavlin** [comment]
Fixed by #7464.
